kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/VirtualCluster.java

```diff
@@ -5,20 +5,80 @@
 
+import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import io.kroxylicious.proxy.config.tls.Tls;
 
 import edu.umd.cs.findbugs.annotations.Nullable;
 
+/**
+ * A virtual cluster.
+ *
+ * @param targetCluster the cluster being proxied
+ * @param clusterNetworkAddressConfigProvider virtual cluster network config - deprecated - use a named listener
+ * @param tls deprecated - tls settings for the virtual cluster - deprecated - use a named listener
+ * @param listeners virtual cluster listeners
+ * @param logNetwork if true, network will be logged
+ * @param logFrames if true, kafka rpcs will be logged
+ * @param filters filers.
+ */
+@SuppressWarnings("java:S1123") // suppressing the spurious warning about missing @deprecated in javadoc. It is the field that is deprecated, not the class.
 public record VirtualCluster(TargetCluster targetCluster,
-                             @JsonProperty(required = true) ClusterNetworkAddressConfigProviderDefinition clusterNetworkAddressConfigProvider,
-                             @JsonProperty() Optional<Tls> tls,
+                             @Deprecated(forRemoval = true, since = "0.11.0") ClusterNetworkAddressConfigProviderDefinition clusterNetworkAddressConfigProvider,
+                             @Deprecated(forRemoval = true, since = "0.11.0") @JsonProperty() Optional<Tls> tls,
+
+                             @JsonProperty(required = false) List<VirtualClusterListener> listeners,
                              boolean logNetwork,
                              boolean logFrames,
                              @Nullable List<String> filters) {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(VirtualCluster.class);
+
+    /**
+     * Name given to the listener defined using the deprecated fields.
+     */
+    @Deprecated(forRemoval = true, since = "0.11.0")
+    static final String DEFAULT_LISTENER_NAME = "default";
+
+    @SuppressWarnings({ "removal", "java:S2789" }) // S2789 - checking for null tls is the intent
+    public VirtualCluster {
+        if (clusterNetworkAddressConfigProvider != null || (tls != null && tls.isPresent())) {
+            if (clusterNetworkAddressConfigProvider == null) {
+                throw new IllegalConfigurationException("Deprecated virtualCluster property 'tls' supplied, but 'clusterNetworkAddressConfigProvider' is null");
+            }
+            if (listeners == null || listeners.isEmpty()) {
+                LOGGER.warn(
+                        "The virtualCluster properties 'clusterNetworkAddressConfigProvider' and 'tls' are deprecated, specify virtual cluster listeners using the listeners map.");
+                listeners = List.of(new VirtualClusterListener(DEFAULT_LISTENER_NAME, clusterNetworkAddressConfigProvider, tls));
```

Comment:
This breaks the YAML fidelity property we have elsewhere, where `yaml == serialize(deserialize(yaml))`. legacy VirtualCluster usages will emit a `listeners` list when you serialize them. This pushed us to update the operator code as part of this PR (not necessarily a bad thing to do, but it does make the PR bigger). The fidelity prop is useful if we want to programmatically construct old YAML for tests etc.

To keep fidelity we are pushed towards something a bit more awkward where the input properties aren't altered, I'm not sure what I prefer in the end.

Reply:
I'll have a go at refactoring the code to keep fidelity.  I think I can do it in a straight-forward way.